dotnetapp-dev/Dockerfile

```diff
@@ -1,11 +1,35 @@
-FROM microsoft/dotnet:2.0-sdk
+FROM microsoft/dotnet:2.0-sdk AS test
```

Comment:
1. Labeling the first stage `test` is confusing as it does nothing related to testing other than it contains the tests.
2. What do you think about performing the build in the first phase.  The benefit of this is that if you run the tests multiple times, you won't incur the overhead of building each time, additionally the publish step can make use of the built code.

The following illustrates these fixes.
Note:  I am on the fence about separating the `test` phase - it supports running only the tests as part of the build without incurring the overhead of publishing.

```
FROM microsoft/dotnet:2.0-sdk AS build
WORKDIR /app

# copy csproj and restore as distinct layers
COPY *.sln .
COPY dotnetapp/*.csproj ./dotnetapp/
COPY botutils/*.csproj ./botutils/
COPY bottests/*.csproj ./bottests/
RUN dotnet restore

# copy and build everything else
COPY dotnetapp/. ./dotnetapp/
COPY botutils/. ./botutils/
COPY bottests/. ./bottests/

RUN dotnet build


FROM build AS testrunner
WORKDIR /app/bottests
ENTRYPOINT ["dotnet", "test","--logger:trx"]


FROM build AS test
WORKDIR /app/bottests
RUN dotnet test


FROM test AS publish
WORKDIR /app/dotnetapp
RUN dotnet publish -o out


FROM microsoft/dotnet:2.0-runtime AS runtime
WORKDIR /app
COPY --from=publish /app/dotnetapp/out ./
ENTRYPOINT ["dotnet", "dotnetapp.dll"]
```